src/Exception/BatchException.php

```diff
@@ -0,0 +1,110 @@
+<?php
+
+namespace Http\Client\Exception;
+
+use Psr\Http\Message\ResponseInterface;
+
+/**
+ * @author GeLo <geloen.eric@gmail.com>
+ */
+final class BatchException extends TransferException
+{
+    /**
+     * @var TransferException[]
+     */
+    private $exceptions;
+
+    /**
+     * @var ResponseInterface[]
+     */
+    private $responses;
+
+    /**
+     * @param TransferException[] $exceptions
+     * @param ResponseInterface[]   $responses
+     */
+    public function __construct(array $exceptions = [], array $responses = [])
+    {
+        parent::__construct('An error occurred when sending multiple requests.');
+
+        foreach ($exceptions as $e) {
+            if (!$e instanceof TransferException) {
+                throw new InvalidArgumentException('Exception is not an instanceof Http\Client\Exception\TransferException');
+            }
+        }
+
+        foreach ($responses as $response) {
+            if (!$response instanceof ResponseInterface) {
+                throw new InvalidArgumentException('Response is not an instanceof Psr\Http\Message\ResponseInterface');
+            }
+        }
+
+
+        $this->exceptions = $exceptions;
+        $this->responses = $responses;
+    }
+
+    /**
+     * Returns all exceptions
+     *
+     * @return TransferException[]
+     */
+    public function getExceptions()
+    {
+        return $this->exceptions;
+    }
+
+    /**
+     * Checks if a specific exception exists
+     *
+     * @param TransferException $exception
+     *
+     * @return boolean TRUE if there is the exception else FALSE.
+     */
+    public function hasException(TransferException $exception)
+    {
+        return array_search($exception, $this->exceptions, true) !== false;
+    }
+
+    /**
+     * Checks if any exception exists
+     *
+     * @return boolean
+     */
+    public function hasExceptions()
+    {
+        return !empty($this->exceptions);
+    }
+
+    /**
+     * Returns all responses
```

Comment:
Another possibility is to store exceptions in an SplObjectStorage indexed by the request objects. This would be the best way to identify every exceptions. I am also thinking about adding a BatchResponse class as well, which also holds an SplObjectStorage internally, which would contain responses indexed by requests. WDYT?